lucille-core/src/main/java/com/kmwllc/lucille/stage/PythonStage.java

```diff
@@ -0,0 +1,379 @@
+      } catch (IOException e) {
+        log.error("port {} is already in use", port, e);
+        portAvailable = false;
+      }
+      if (!portAvailable) {
+        log.warn(
+            "WARNING: Port {} is already in use. Assuming a GatewayServer is already running and reconnecting to it.",
+            port);
+        // Do not start a new GatewayServer or Python process
+        gateway = null;
+        pythonProcess = null;
+        return;
+      }
+
+      log.info("Starting Py4J GatewayServer on port {}", port);
+      gateway = new GatewayServer(this, port);
+      gateway.start();
+      log.info("GatewayServer started");
+
+      handler = (Py4jExecutorInterface) gateway.getPythonServerEntryPoint(new Class[] { Py4jExecutorInterface.class });
+
+      try {
+        // Use Py4jClient.py from main/resources
+        String py4jClientScript = getClass().getClassLoader().getResource("Py4jClient.py").getPath();
+        log.info("Using Py4jClient.py script at {}", py4jClientScript);
+
+        log.info("Launching Python process: {} {} {} {}", pythonExecutable, py4jClientScript, scriptPath, port);
+        pythonProcess = new ProcessBuilder(
+                pythonExecutable,
+                py4jClientScript,
+                scriptPath,
+                String.valueOf(port)
+            )
+            .redirectErrorStream(true)
+            .start();
+        log.info("Python process started");
+
+        py4jClient = new Py4jClient(this, pythonProcess);
+
+      } catch (Exception e) {
+        log.error("Failed to launch Python process for Py4J integration", e);
+        throw new StageException("Failed to launch Python process for Py4J integration", e);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    log.info("processDocument");
+    try {
+      Map<String, Object> msg = new HashMap<>();
+      msg.put("method", "process_document");
+      msg.put("data", doc);
+      String msgJson = mapper.writeValueAsString(msg);
+      log.info("JSON encoded doc: {}", msgJson);
+      // Call Python and get the updated document as a JSON string
+      String jsonReturn = (String) handler.send(msgJson); // Assumes send returns the result
+      if (jsonReturn != null) {
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> updatedFields = mapper.readValue(jsonReturn, Map.class);
+
+        Set<String> reserved = com.kmwllc.lucille.core.Document.RESERVED_FIELDS;
+        for (String field : new HashSet<>(doc.getFieldNames())) {
+          if (!reserved.contains(field)) {
+            doc.removeField(field);
+          }
+        }
+
+        for (Map.Entry<String, Object> entry : updatedFields.entrySet()) {
+          if (!reserved.contains(entry.getKey())) {
+            doc.setField(entry.getKey(), entry.getValue());
+          }
+        }
+      }
+      doc.setField("python_stage_executed", true);
+      return null;
```

Comment:
normally , this return value is an iterator to any children documents returned by the stage.   this implies that this stage can not emit children documents.